saleor/checkout/complete_checkout.py

```diff
@@ -342,6 +343,8 @@ def _create_order(
     """
     from ..order.utils import add_gift_card_to_order
 
+    Allocation.objects.select_for_update(of=("self",))
```

Comment:
Have you tried locking allocation inside `allocate_stocks`? This is the function that is supposed to handle that, so it would be nice to have this logic there.

Reply:
Yes, I tried but without any results. Only adding this lock at the beginning of completing checkout works fine.

Reply:
Are you sure of this? I see that `allocate_stocks` raises the exception when there are not enough items in the stock.
Also, you're allocating here all allocations which means that you block the possibility to process more than one checkoutComplete at the same time (AFAIK).

Reply:
Yes, I'm sure. I added this lock into allocate_stocks and that doesn't prevent this negative quantity

Reply:
But if you add lock in this way, it's lock read of any allocation(Any client can't check available quantity) when the server run  `_create_order` for any order.

Reply:
imo, it is workaround for something that doesn't work as we expect. Locking all allocations will force Saleor to **process all checkouts one by one**, which is not the best way for handling this case. 
Have you investigated which place in code causes the issue?